### ISSUE TITLE
Multiple video streams for the easyrtc adapter

### ISSUE THREAD

**Issue by vincentfretin** [networked-aframe/networked-aframe#269]
We added camera streaming in networted-aframe 0.8.0 with #261 and a fix just after in 0.8.2 https://github.com/networked-aframe/networked-aframe/commit/2ae6b3b8eda74091452ff7bdeb84ab4f2d890a4e

It would be great if we could be able to easily add screen sharing or an arbitrary video stream with camera on, so two video tracks or more.
On Open-EasyRTC side you can register a new stream using the `easyrtc.register3rdPartyLocalMediaStream` API.
```
navigator.mediaDevices.getDisplayMedia({video: true}).then(function (stream) {
    easyrtc.register3rdPartyLocalMediaStream(stream, 'screen:1');

    // Then to add to existing connection
    easyrtc.addStreamToCall(easyrtcId, 'screen:1', function (caller, streamName) {

    });
});
```

The janus adapter has an API `setLocalMediaStream(stream)` (stream includes zero or one audio track and zero or one video track) that you can call several times. It will replace or disable the audio or video track on the RTCPeerConnection, but in janus-plugin-sfu it's limited to one audio track and one video track on the RTCPeerConnection.

We could expand on this API to allow multiple video tracks.
- `addLocalMediaStream(stream, streamName)`
- `removeLocalMediaStream(stream, streamName)`
- `getMediaStream(clientId, streamName)` (currently streamName can only be audio or video, can be now any name?)

There are lots of things to considerate here, the renegotiation of the SDP offer/answer when adding and removing a video track, what to do at the aframe component level when a video stream is ended, and maybe more I didn't think about.

**Comment by vincentfretin** [comment]
@keianhzo @netpro2k for the dialog adapter and @mqp for the janus adapter. If we can agree on the same API for multiple tracks, that would be great. If you have some insights that I didn't think about, please share.

In the janus adapter there is 
- `enableMicrophone(enabled)`

We added it in the easyrtc adapter with:
- `enableCamera(enabled)`

I see in the [dialog adapter](https://github.com/mozilla/hubs/blob/master/src/naf-dialog-adapter.js) you have:
- `async enabledMic(track)`
- `async enableCamera(track)`
- `async disableCamera()`
- `async enableShare(track)`
- `async disableShare()`
- `enableMicrophone(enabled)`

**Comment by jimver04** [comment]
I have done some checks here and I think this is an open-easyrtc issue. I have opened an issue there:

https://github.com/open-easyrtc/open-easyrtc/issues/73

It is a pity that video streaming is limited only to camera. I tried to replace the camera stream with canvasCtx.captureStream(myFrameRate) but no luck, I keep getting my video stream instead of the other player stream.

**Comment by hthetiot** [comment]
It's not a open-easyrtc issue, since easyrtc does support multiple streams.

All you need to do is to use `easyrtc.register3rdPartyLocalMediaStream` with capture from canvas and `easyrtc.addStreamToCall` if the stream is added after connexion.

**Comment by jimver04** [comment]
I just want to do some video effects on my camera and send it. I do not want to increase the bandwidth by sending two streams. I guess I have to 

1) register new custom stream, e.g. 'FXcamera' (easyrtc.register3rdPartyLocalMediaStream with capture from canvas  )
2) remove the normal video stream from the call 
3) add new stream to call (easyrtc.addStreamToCall)
4) set getMediaStream to take my own 'FXcamera' stream ?

and that's all ???

**Comment by hthetiot** [comment]
Yes mostly, best is to extend EasyRtcAdapter to MyEasyRtcAdapter and override _connect to replace call to initMediaSource by register3rdPartyLocalMediaStream with your canvas, like that no addStreamToCall is needed.
This issue was originally created to align the EasyRtcAdapter Api to implement addLocalMediaStream/removeLocalMediaStream to ease multi stream but i dont have time to do that for now and I would personally suggest to create a separate issue in the future even if you case fit in the removeLocalMediaStream/removeLocalMediaStream future use-case but is not narrowed to Canvas Capture.

**Comment by jimver04** [comment]
I have spent 5 hours today, but no lack. I am getting the other player video stream instead of other player screen1. How do I retrieve the "screen1" stream ?

This is what I have in my example

```
NAF.utils.getNetworkedEntity(this.el).then((networkedEl) => {
                const ownerId = networkedEl.components.networked.data.owner;
                if (ownerId) {
                    NAF.connection.adapter.getMediaStream(ownerId, "screen1")
                        .then(this._setMediaStream)
                        .catch((e) => NAF.log.error(`Error getting media stream for ${ownerId}`, e));
                } else {
                    // Correctly configured local entity, perhaps do something here for enabling debug audio loopback
                }
            });
```


And these are the changes in EasyRtcAdapter.js. I have not commented  this.easyrtc.initMediaSource because there are no streams in the receiver if I comment it.


 ```
   getMediaStream(clientId, type = "audio") {
    if (this.mediaStreams[clientId]) {
      NAF.log.write(`Already had ${type} for ${clientId}`);
      return Promise.resolve(this.mediaStreams[clientId][type]);
    } else {
      NAF.log.write(`Waiting on ${type} for ${clientId}`);
      if (!this.pendingMediaRequests.has(clientId)) {
        this.pendingMediaRequests.set(clientId, {});

        const audioPromise = new Promise((resolve, reject) => {
          this.pendingMediaRequests.get(clientId).audio = { resolve, reject };
        });
        const videoPromise = new Promise((resolve, reject) => {
          this.pendingMediaRequests.get(clientId).video = { resolve, reject };
        });
        const screen1Promise = new Promise((resolve, reject) => {
          this.pendingMediaRequests.get(clientId).screen1 = { resolve, reject };
        });


        this.pendingMediaRequests.get(clientId).audio.promise = audioPromise;
        this.pendingMediaRequests.get(clientId).video.promise = videoPromise;
        this.pendingMediaRequests.get(clientId).screen1.promise = screen1Promise;

        audioPromise.catch(e => NAF.log.warn(`${clientId} getMediaStream Audio Error`, e));
        videoPromise.catch(e => NAF.log.warn(`${clientId} getMediaStream Video Error`, e));
        screen1Promise.catch(e => NAF.log.warn(`${clientId} getMediaStream Screen1 Error`, e));
      }
      return this.pendingMediaRequests.get(clientId)[type].promise;
    }
  }

  setMediaStream(clientId, stream) {
    // Safari doesn't like it when you use single a mixed media stream where one of the tracks is inactive, so we
    // split the tracks into two streams.
    const audioStream = new MediaStream();
    try {
      stream.getAudioTracks().forEach(track => audioStream.addTrack(track));
    } catch(e) {
      NAF.log.warn(`${clientId} setMediaStream Audio Error`, e);
    }

    const videoStream = new MediaStream();
    try {
      stream.getVideoTracks().forEach(track => videoStream.addTrack(track));
    } catch (e) {
      NAF.log.warn(`${clientId} setMediaStream Video Error`, e);
    }

    const screen1Stream = new MediaStream();
    try {
      stream.getVideoTracks().forEach(track => screen1Stream.addTrack(track));
    } catch (e) {
      NAF.log.warn(`${clientId} setMediaStream Screen1 Error`, e);
    }

    this.mediaStreams[clientId] = { audio: audioStream, video: videoStream, screen1: screen1Stream };

    // Resolve the promise for the user's media stream if it exists.
    if (this.pendingMediaRequests.has(clientId)) {
      this.pendingMediaRequests.get(clientId).audio.resolve(audioStream);
      this.pendingMediaRequests.get(clientId).video.resolve(videoStream);
      this.pendingMediaRequests.get(clientId).screen1.resolve(screen1Stream);
    }
  }

  enableMicrophone(enabled) {
    this.easyrtc.enableMicrophone(enabled);
  }

  enableCamera(enabled) {
    this.easyrtc.enableCamera(enabled);
  }

  disconnect() {
    this.easyrtc.disconnect();
  }

  /**
   * Privates
   */

  _connect(connectSuccess, connectFailure) {
    var that = this;

    this.easyrtc.setStreamAcceptor(this.setMediaStream.bind(this));

    this.easyrtc.setOnStreamClosed(function(easyrtcid) {
      delete that.mediaStreams[easyrtcid];
    });

    this.easyrtc.register3rdPartyLocalMediaStream(canvasStream, "screen1");


    if (that.easyrtc.audioEnabled || that.easyrtc.videoEnabled) {
      this.easyrtc.initMediaSource(
        function(stream) {
          that.setMediaStream(that.easyrtc.myEasyrtcid, stream);
          that.easyrtc.connect(that.app, connectSuccess, connectFailure);
        },
        function(errorCode, errmesg) {
          NAF.log.error(errorCode, errmesg);
        }
      );
    } else {
      that.easyrtc.connect(that.app, connectSuccess, connectFailure);
    }
  }
```

**Comment by vincentfretin** [comment]
In your `setMediaStream` function, I think you end up having the two video tracks in `videoStream` and in `screen1Stream` because you don't do any filter. `video.srcObject` in your modified `networked-video-source` component will take I think the first video track in the screen1Stream stream, which is probably the camera. You need to add a filter, maybe look at the `track.name`.

**Comment by jimver04** [comment]
There is only one video track in the stream and this is the camera but not my canvas stream. It seems that register3rdPartyLocalMediaStream   should be placed somewhere else. It just does this

`namedLocalMediaStreams[streamName] = stream;`

```
    this.register3rdPartyLocalMediaStream = function(stream, streamName) {
       return registerLocalMediaStreamByName(stream, streamName);
    };

/** @private */
    function registerLocalMediaStreamByName(stream, streamName) {
        var roomName;
        if (!streamName) {
            streamName = "default";
        }
        stream.streamName = streamName;
        namedLocalMediaStreams[streamName] = stream;
        if (streamName !== "default") {
            var mediaIds = buildMediaIds(),
                roomData = self.roomData;
            for (roomName in roomData) {
                if (roomData.hasOwnProperty(roomName)) {
                    self.setRoomApiField(roomName, "mediaIds", mediaIds);
                }
            }
        }
    }
```

register3rdPartyLocalMediaStream    does not seem to add new video tracks to the existing stream but a brand new stream

**Comment by jimver04** [comment]
Oh my god it was that. No necessary to use register3rdpartystream at all
just add an extra stream inside init "

```
if (that.easyrtc.audioEnabled || that.easyrtc.videoEnabled) {
      this.easyrtc.initMediaSource(
        function(stream) {
          //that.easyrtc.register3rdPartyLocalMediaStream(canvasStream, "screen1");

          stream.addTrack(canvasStream.getVideoTracks()[0]);


          that.setMediaStream(that.easyrtc.myEasyrtcid, stream);

          that.easyrtc.connect(that.app, connectSuccess, connectFailure);
        },
        function(errorCode, errmesg) {
          NAF.log.error(errorCode, errmesg);
        }
      );
    } else {
      that.easyrtc.connect(that.app, connectSuccess, connectFailure);
    }
```

and when you want to use that additional stream as a texture do this:

 ```
let mMs = new MediaStream();
 mMs.addTrack(newStream.getVideoTracks()[1])
 this.video.srcObject = mMs;
 this.videoTexture = new THREE.VideoTexture(this.video);
```

notice that I get the second stream ()[1] which is my canvas generated stream. Now I have to remove track 0 to save bandwidth before connection but this would be easy , I guess. Ok, the bacgkround is black (no transparent pixels are transferred, however this can be solved with custom shaders). Are you interested to send this as a new EasyRTCAdapter and a new example in NAF ?

 Best,
Dimitrios

**Comment by hthetiot** [comment]
I do not recommend to add more than one track of a given type in a media stream, it may work today in your environment but will not work necessarily in the future, it may also have side unexpected side effects. Also It will work prior connection but not to my knowledge if peer connection is already established since new track will not be taken into account if peer connection is established since RTCsender will not be created or updated.

 No to mention that you may end up with dead RTCsenders when the track is stopped.

**Comment by hthetiot** [comment]
> register3rdPartyLocalMediaStream does not seem to add new video tracks to the existing stream but a brand new stream

Yes it create a new stream that the expected behavior and will trigger hook setStreamAcceptor on remote peer.

**Comment by hthetiot** [comment]
> Now I have to remove track 0 to save bandwidth before connection but this would be easy

Remove track without renegotiate will result in RTCsender with dead track that not how you should manage tracks.

**Comment by hthetiot** [comment]
> Are you interested to send this as a new EasyRTCAdapter and a new example in NAF ?

No sorry because this is not the way it should be done. But thank you.
This is basically a hack for a single stream with multiple video tracks.

Following @vincentfretin instructions using `register3rdPartyLocalMediaStream`, `addStreamToCall`  and implementing `addLocalMediaStream`, `removeLocalMediaStream `  would have resulted in perfectly fine implementation, see this PR with functional,multi-video stream example.
- https://github.com/networked-aframe/networked-aframe/pull/294

The part you missed is the capability of informing networked-video-source component what stream it need to use.

```
<a-plane networked-video-source></a-plane>
<a-plane networked-video-source="streamName: screen"></a-plane>
````

<img width="1429" alt="Screenshot 2021-09-13 at 18 15 54" src="https://user-images.githubusercontent.com/8635/133120070-08a62806-c07e-4390-9a92-67bfcc5cb92a.png">

**Comment by hthetiot** [comment]
> There are lots of things to considerate here, the renegotiation of the SDP offer/answer when adding and removing a video track

This part is handled by easyRTC.

> what to do at the aframe component level when a video stream is ended, and maybe more I didn't think about.
This part I don't really know how to implement, I would prefer you provide guidance or build upon the new multi-stream API. One solution would be to only display video component if stream is started.

**Comment by vincentfretin** [comment]
Thanks @hthetiot I'll take a look at your PR in depth soon and test it, but not today, probably at the end of the week.

>> what to do at the aframe component level when a video stream is ended, and maybe more I didn't think about.
> 
> This part I don't really know how to implement, I would prefer you provide guidance or build upon the new multi-stream API. One solution would be to only display video component if stream is started.

I'm not sure if this works correctly on the receiving part, we may use `newStream.getVideoTracks()[0].onended=() => this.el.setAttribute('visible", false)` (completely not tested) to hide the entity if it stops. But it would be nice to show again the entity if the state change again to something different than ended. If the user stops and restarts their video, do we have the same stream and track or do we need to get again the media? I didn't look deeply at the code yet, sorry if I'm asking something obvious. :)
If we can detect somehow that the stream is playing and change the entity visibility or change the plane color to grey or whatever (maybe be able to configure a callback when the stream state change, with a default callback setting the plane visibility), that would be great.

**Comment by hthetiot** [comment]
> I'm not sure if this works correctly on the receiving part, we may use newStream.getVideoTracks()[0].onended=() => this.el.setAttribute('visible", false) (completely not tested) to hide the entity if it stops. But it would be nice to show again the entity if the state change again to something different than ended. If the user stops and restarts their video, do we have the same stream and track or do we need to get again the media? I didn't look deeply at the code yet, sorry if I'm asking something obvious. :)
If we can detect somehow that the stream is playing and change the entity visibility or change the plane color to grey or whatever (maybe be able to configure a callback when the stream state change, with a default callback setting the plane visibility), that would be great.

I will make separate PR for that, i have good idea how to handle this state.

**Comment by vincentfretin** [comment]
@hthetiot I closed this issue and created #299 for the video track stopped and restarted issue.

**Comment by hthetiot** [comment]
Thank you @vincentfretin i did not have time but you did great.
